### PR TITLE
r/codebuild_project: Support type for environment_variable

### PR DESCRIPTION
**Pull request by atsushi-ishibashi**
Supported: #1747

- [x] acceptance test

### CODE REVIEW

**Comment by katgironpe** [comment]
LGTM. Please merge!

File: website/docs/r/codebuild_project.html.markdown [hunk 1]
```diff
@@ -134,6 +134,7 @@ The following arguments are supported:
 `environment_variable` supports the following:
```
Comment (nicwise) [inline website/docs/r/codebuild_project.html.markdown]: Any chance you can add an extra CR at the end of this? the docs don't parse properly otherwise.

**Comment by nicwise** [comment]
Please Do! Just one change to the docs first tho (optionally)

website/docs/r/codebuild_project.html.markdown

File: aws/resource_aws_codebuild_project.go [hunk 2]
```diff
@@ -95,6 +95,21 @@ func resourceAwsCodeBuildProject() *schema.Resource {
...
+											value := v.(string)
+											validType := []string{"PARAMETER_STORE", "PLAINTEXT"}
+											for _, str := range validType {
+												if value == str {
+													return
+												}
+											}
+											errors = append(errors, fmt.Errorf("expected %s to be one of %v, got %s", k, validType, value))
+											return
+										},
```
Comment (radeksimko) [inline aws/resource_aws_codebuild_project.go]: Nitpick: Do you mind replacing this with a standard helper we have in the upstream library and constants from the SDK? i.e.

```go
ValidateFunc: validation.StringInSlice([]string{
  codebuild.EnvironmentVariableTypeParameterStore,
  codebuild.EnvironmentVariableTypePlaintext,
}, true),
```

**Comment by nicwise** [comment]
@radeksimko Any chance this can go into the next release? (and thank you @atsushi-ishibashi !)

**Comment by nicwise** [comment]
(incase @atsushi-ishibashi doesn't have time to fix the doc thing, I did a PR with just that in, and his change: https://github.com/terraform-providers/terraform-provider-aws/pull/3323 )

**Comment by bflad** [comment]
`environmentVariablesToMap` will likely also need an update so the new optional attribute properly gets read (the missing acceptance test would've caught this):

```go
if env.Type != nil {
  item["type"] = *env.Type
}
```

**Comment by atsushi-ishibashi** [comment]
```
TF_ACC=1 go test ./aws -v -run=TestAccAWSCodeBuildProject_basic -timeout 120m
=== RUN   TestAccAWSCodeBuildProject_basic
--- PASS: TestAccAWSCodeBuildProject_basic (96.77s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	96.810s
```

File: website/docs/r/codebuild_project.html.markdown (hunk 1 above)
Comment (nicwise) [reply website/docs/r/codebuild_project.html.markdown]: This is done and merged BTW

**Comment by nicwise** [comment]
@bflad / @catsby - looks like @atsushi-ishibashi has added the `environmentVariablesToMap` thing - is this one good to go? ;-)

**Comment by bflad** [comment]
For the acceptance testing, we'd normally expect a `_basic` test to not include an optional attribute (duplicating the entire test and config is fine) and for the test including a new attribute to at least check the attribute value in the Terraform state via `resource.TestCheckResourceAttr(resourceName, "attribute", "expectedValue")`.

If you don't have time to submit these fixes, please let us know and we can finish them up. 👍

**Comment by larsjarek** [comment]
Hey guys, is there a chance that this feature can go into the next release? I would really appreciate it as it is needed for my AWS CodeBuild configuration. If @atsushi-ishibashi does not have time I might fix these tests in another PR.

**Comment by larsjarek** [comment]
I made a PR with fix for tests. Currently this PR is marked as WIP because I have to resolve some conflicts.

**Comment by katgironpe** [comment]
@larsjarek Hey thanks for this. Looking forward to that getting merged.

**Comment by atsushi-ishibashi** [comment]
Sorry for replying late. I have no time to update so I'm glad if anyone take this PR over:bow:

**Comment by larsjarek** [comment]
No problem @atsushi-ishibashi, thanks for your contribution up to this point. All conflicts in my PR (#4021) are now resolved, it can be reviewed now.

**Comment by bnightstar** [comment]
when this will be merged I need this for a project I'm working on ?

**Comment by larsjarek** [comment]
It does not seem that this feature will be merged soon.

**Comment by bflad** [comment]
The later #4021 (containing this PR) is actively being reviewed this week.

**Comment by bflad** [comment]
After looking through the various duplicate pull requests, I'm going to try to get the commits merged in the following order here (resolving the merge conflicts and test changes along the way, which could be tedious): #2811 -> #4021 -> #3929. 🤞 

In the end the following new arguments will be supported:
* `environment` > `environment_variable` > `type`
* `source` > `git_clone_depth`
* `source` > `insecure_ssl`

**Comment by bflad** [comment]
After merging together the above pull requests and refactoring the test suite, we should be good to go:

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSCodeBuildProject'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSCodeBuildProject -timeout 120m
=== RUN   TestAccAWSCodeBuildProject_basic
--- PASS: TestAccAWSCodeBuildProject_basic (31.87s)
=== RUN   TestAccAWSCodeBuildProject_BadgeEnabled
--- PASS: TestAccAWSCodeBuildProject_BadgeEnabled (30.05s)
=== RUN   TestAccAWSCodeBuildProject_BuildTimeout
--- PASS: TestAccAWSCodeBuildProject_BuildTimeout (38.35s)
=== RUN   TestAccAWSCodeBuildProject_Cache
--- PASS: TestAccAWSCodeBuildProject_Cache (59.77s)
=== RUN   TestAccAWSCodeBuildProject_Description
--- PASS: TestAccAWSCodeBuildProject_Description (38.08s)
=== RUN   TestAccAWSCodeBuildProject_EncryptionKey
--- PASS: TestAccAWSCodeBuildProject_EncryptionKey (53.94s)
=== RUN   TestAccAWSCodeBuildProject_Environment_EnvironmentVariable_Type
--- PASS: TestAccAWSCodeBuildProject_Environment_EnvironmentVariable_Type (30.31s)
=== RUN   TestAccAWSCodeBuildProject_Source_Auth
--- PASS: TestAccAWSCodeBuildProject_Source_Auth (20.33s)
=== RUN   TestAccAWSCodeBuildProject_Source_GitCloneDepth
--- PASS: TestAccAWSCodeBuildProject_Source_GitCloneDepth (38.51s)
=== RUN   TestAccAWSCodeBuildProject_Source_InsecureSSL
--- PASS: TestAccAWSCodeBuildProject_Source_InsecureSSL (37.43s)
=== RUN   TestAccAWSCodeBuildProject_Tags
--- PASS: TestAccAWSCodeBuildProject_Tags (28.91s)
=== RUN   TestAccAWSCodeBuildProject_VpcConfig
--- PASS: TestAccAWSCodeBuildProject_VpcConfig (57.63s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	465.209s
```

This will release with v1.21.0 of the AWS provider, likely middle of next week.

**Comment by larsjarek** [comment]
@bflad Thanks a lot for merging this. Is there a date when new version which includes this functionality will be released?

**Comment by bflad** [comment]
This has been released in [version 1.21.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1210-may-31-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**